util/kvencoder/kv_encoder_test.go

```diff
@@ -0,0 +1,383 @@
+			c.Assert(emptyValCount, Equals, ca.expectEmptyValCnt, Commentf(comment))
+		}
+	}
+}
+
+func (s *testKvEncoderSuite) TestInsertPkIsHandle(c *C) {
+	defer s.cleanEnv(c)
+	tk := testkit.NewTestKit(c, s.store)
+	tkExpect := testkit.NewTestKit(c, s.storeExpect)
+	tk.MustExec("use test")
+	tkExpect.MustExec("use test")
+
+	encoder, err := New(s.store, "test", nil, true)
+	c.Assert(err, IsNil)
+
+	schemaSQL := "create table t(id int auto_increment, a char(10), primary key(id))"
+	tk.MustExec(schemaSQL)
+	tkExpect.MustExec(schemaSQL)
+
+	sqls := []testCase{
+		{"insert into t values(1, 'test');", 0, 1},
+		{"insert into t(a) values('test')", 0, 1},
+		{"insert into t(id, a) values(3, 'test')", 0, 1},
+		{"insert into t values(1000000, 'test')", 0, 1},
+		{"insert into t(a) values('test')", 0, 1},
+		{"insert into t(id, a) values(4, 'test')", 0, 1},
+	}
+
+	s.runTestSQL(c, tk, tkExpect, encoder, sqls)
+
+	schemaSQL = "create table t1(id int auto_increment, a char(10), primary key(id), key a_idx(a))"
+	tk.MustExec(schemaSQL)
+	tkExpect.MustExec(schemaSQL)
+
+	sqls = []testCase{
+		{"insert into t1 values(1, 'test');", 0, 2},
+		{"insert into t1(a) values('test')", 0, 2},
+		{"insert into t1(id, a) values(3, 'test')", 0, 2},
+		{"insert into t1 values(1000000, 'test')", 0, 2},
+		{"insert into t1(a) values('test')", 0, 2},
+		{"insert into t1(id, a) values(4, 'test')", 0, 2},
+	}
+
+	s.runTestSQL(c, tk, tkExpect, encoder, sqls)
+
+	schemaSQL = `create table t2(
+		id int auto_increment, 
+		a char(10), 
+		b datetime default NULL, 
+		primary key(id), 
+		key a_idx(a),
+		unique b_idx(b))
+		`
+
+	tk.MustExec(schemaSQL)
+	tkExpect.MustExec(schemaSQL)
+
+	sqls = []testCase{
+		{"insert into t2(id, a) values(1, 'test')", 0, 3},
+		{"insert into t2 values(2, 'test', '2017-11-27 23:59:59')", 0, 3},
+		{"insert into t2 values(3, 'test', '2017-11-28 23:59:59')", 0, 3},
+	}
+	s.runTestSQL(c, tk, tkExpect, encoder, sqls)
+}
+
+func (s *testKvEncoderSuite) TestInsertPkIsNotHandle(c *C) {
+	defer s.cleanEnv(c)
+	tk := testkit.NewTestKit(c, s.store)
+	tkExpect := testkit.NewTestKit(c, s.storeExpect)
+	tk.MustExec("use test")
+	tkExpect.MustExec("use test")
+
+	encoder, err := New(s.store, "test", nil, true)
+	c.Assert(err, IsNil)
+
+	schemaSQL := `create table t(
+		id varchar(20),
+		a char(10),
+		primary key(id))`
+	tk.MustExec(schemaSQL)
+	tkExpect.MustExec(schemaSQL)
+
+	sqls := []testCase{
+		{"insert into t values(1, 'test');", 0, 2},
+		{"insert into t(id, a) values(2, 'test')", 0, 2},
+		{"insert into t(id, a) values(3, 'test')", 0, 2},
+		{"insert into t values(1000000, 'test')", 0, 2},
+		{"insert into t(id, a) values(5, 'test')", 0, 2},
+		{"insert into t(id, a) values(4, 'test')", 0, 2},
+	}
+
+	s.runTestSQL(c, tk, tkExpect, encoder, sqls)
+}
+
+func insertData(tk, tkExpect *testkit.TestKit, sqls []string) {
+	for _, sql := range sqls {
+		tk.MustExec(sql)
+		tkExpect.MustExec(sql)
+	}
+}
+
+func (s *testKvEncoderSuite) TestUpdate(c *C) {
```

Comment:
I thought KVEncoder is only used for INSERT.

Reply:
But actually, just don't want this api to be specific.